File: apps/android-client/src/LocalRpcCoreBridgeClient.kt
package com.gogate.android

import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.concurrent.thread

class LocalRpcCoreBridgeClient(
    private val baseUrl: String = "http://127.0.0.1:47831",
    private val pollIntervalMs: Long = 750L,
    private val maxBackoffMs: Long = 8000L,
    private val failureThresholdForDegraded: Int = 2
) : CoreBridgeClient, AutoCloseable {

    private val listeners = ConcurrentHashMap<String, (BridgeEvent) -> Unit>()
    @Volatile private var polling = false
    @Volatile private var pollFailureCount = 0
    @Volatile private var currentBackoffMs = pollIntervalMs
    @Volatile private var bridgeHealthy = true
    @Volatile private var lastBridgeReason = "ok"
    @Volatile private var lastEventFingerprint = ""

    override fun connect(request: ConnectRequest): ConnectResponse {
        val payload = JSONObject()
            .put("profile_id", request.profileId)
            .put("mode", request.mode)
        val data = sendCommand("connect", payload, "connect")
        return if (data != null) {
            ConnectResponse(data.optString("session_id", ""), data.optString("state", "failed"))
        } else {
            ConnectResponse("", "failed")
        }
    }

    override fun disconnect(request: DisconnectRequest): DisconnectResponse {
        val payload = JSONObject().put("session_id", request.sessionId)
        val data = sendCommand("disconnect", payload, "disconnect")
        return DisconnectResponse(data?.optString("state", "failed") ?: "failed")
    }

    override fun rankCandidates(mode: String): RankCandidatesResponse {
        val payload = JSONObject().put("mode", mode)
        val data = sendCommand("rank_candidates", payload, "rank_candidates")
            ?: return RankCandidatesResponse(emptyList())
        val arr = data.optJSONArray("candidates") ?: JSONArray()
        val out = mutableListOf<RankedCandidate>()
        for (i in 0 until arr.length()) {
            val it = arr.optJSONObject(i) ?: continue
            out += RankedCandidate(
                profileId = it.optString("profile_id", ""),
                score = it.optDouble("score", 0.0),
                reason = it.optString("reason", "")
            )
        }
        return RankCandidatesResponse(out)
    }

    override fun fetchLogs(sessionId: String, level: String): FetchLogsResponse {
        val payload = JSONObject().put("session_id", sessionId).put("level", level)
        val data = sendCommand("fetch_logs", payload, "fetch_logs")
        if (data == null) {
            return FetchLogsResponse(listOf(LogEvent(Instant.now().toString(), level, "log unavailable")))
        }
        val arr = data.optJSONArray("events") ?: JSONArray()
        val out = mutableListOf<LogEvent>()
        for (i in 0 until arr.length()) {
            val it = arr.optJSONObject(i) ?: continue
            out += LogEvent(
                ts = it.optString("ts", Instant.now().toString()),
                level = it.optString("level", level),
                message = it.optString("message", "")
            )
        }
        return FetchLogsResponse(out)
    }

    override fun subscribe(listener: (BridgeEvent) -> Unit): DisposableSubscription {
        val id = UUID.randomUUID().toString()
        listeners[id] = listener
        ensurePolling()

        listener(BridgeEvent("bridge_health_changed", mapOf(
            "healthy" to bridgeHealthy.toString(),
            "reason" to lastBridgeReason
        )))

        return DisposableSubscription {
            listeners.remove(id)
            if (listeners.isEmpty()) polling = false
        }
    }

    private fun ensurePolling() {
        if (polling) return
        polling = true
        thread(start = true, isDaemon = true) {
            while (polling) {
                try {
                    val events = pollEvents()
                    pollFailureCount = 0
                    currentBackoffMs = pollIntervalMs
                    markBridgeHealthy("ok")
                    events.forEach(::emitDedup)
                } catch (e: Exception) {
                    registerPollFailure("poll: ${e::class.simpleName ?: "error"}")
                }
                val delay = if (pollFailureCount > 0) currentBackoffMs else pollIntervalMs
                try {
                    Thread.sleep(delay)
                } catch (_: InterruptedException) {
                    return@thread
                }
            }
        }
    }

    private fun registerPollFailure(reason: String) {
        pollFailureCount += 1
        if (pollFailureCount >= failureThresholdForDegraded) {
            markBridgeUnhealthy(reason)
        }
        val next = if (currentBackoffMs <= 0) pollIntervalMs else currentBackoffMs * 2
        currentBackoffMs = minOf(next, maxBackoffMs)
    }

    private fun markBridgeHealthy(reason: String) {
        if (bridgeHealthy && lastBridgeReason == reason) return
        bridgeHealthy = true
        lastBridgeReason = reason
        emit(BridgeEvent("bridge_health_changed", mapOf("healthy" to "true", "reason" to reason)))
    }

    private fun markBridgeUnhealthy(reason: String) {
        if (!bridgeHealthy && lastBridgeReason == reason) return
        bridgeHealthy = false
        lastBridgeReason = reason
        emit(BridgeEvent("bridge_health_changed", mapOf("healthy" to "false", "reason" to reason)))
    }

    private fun emit(event: BridgeEvent) {
        listeners.values.forEach { it(event) }
    }

    private fun emitDedup(event: BridgeEvent) {
        val fingerprint = event.name + ":" + event.fields.toSortedMap().entries.joinToString(";") { "${it.key}=${it.value}" }
        if (fingerprint == lastEventFingerprint) return
        lastEventFingerprint = fingerprint
        emit(event)
    }

    private fun pollEvents(): List<BridgeEvent> {
        val conn = (URL("$baseUrl/events/poll").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.inputStream.use { input ->
            val body = BufferedReader(InputStreamReader(input)).readText()
            val root = JSONObject(body)
            if (!root.optBoolean("ok", false)) throw IllegalStateException("poll bad response")
            val arr = root.optJSONArray("events") ?: JSONArray()
            val out = mutableListOf<BridgeEvent>()
            for (i in 0 until arr.length()) {
                val o = arr.optJSONObject(i) ?: continue
                val fieldsObj = o.optJSONObject("fields") ?: JSONObject()
                val fields = mutableMapOf<String, String>()
                val keys = fieldsObj.keys()
                while (keys.hasNext()) {
                    val k = keys.next()
                    fields[k] = fieldsObj.optString(k, "")
                }
                out += BridgeEvent(o.optString("name", ""), fields)
            }
            return out
        }
    }

    private fun sendCommand(command: String, payload: JSONObject, opName: String): JSONObject? {
        return try {
            val conn = (URL("$baseUrl/command").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = true
                connectTimeout = 6000
                readTimeout = 6000
                setRequestProperty("Content-Type", "application/json")
            }
            val body = JSONObject()
                .put("command", command)
                .put("payload", payload)
                .toString()
            OutputStreamWriter(conn.outputStream).use { it.write(body) }

            val input = if (conn.responseCode < 400) conn.inputStream else conn.errorStream
            if (input == null) {
                markBridgeUnhealthy("$opName: no-response")
                null
            } else {
                val raw = BufferedReader(InputStreamReader(input)).readText()
                val root = JSONObject(raw)
                if (!root.optBoolean("ok", false)) {
                    markBridgeUnhealthy("$opName: bad-response")
                    null
                } else {
                    markBridgeHealthy("ok")
                    root.optJSONObject("data")
                }
            }
        } catch (e: Exception) {
            markBridgeUnhealthy("$opName: ${e::class.simpleName ?: "error"}")
            null
        }
    }

    override fun close() {
        polling = false
    }
}

File: apps/windows-client/src/LocalRpcCoreBridgeClient.cs
using System.Collections.Concurrent;
using System.Text;
using System.Text.Json;
using System.Linq;

namespace GoGate.WindowsClient;

public class LocalRpcCoreBridgeClient : ICoreBridgeClient, IDisposable
{
    private readonly HttpClient _http;
    private readonly CoreBridgeTransportOptions _opt;
    private readonly ConcurrentDictionary<Guid, Action<BridgeEvent>> _subs = new();
    private CancellationTokenSource? _pollCts;
    private Task? _pollTask;

    private int _pollFailureCount;
    private int _currentBackoffMs;
    private bool _bridgeHealthy = true;
    private string _lastBridgeReason = "ok";
    private string _lastEventFingerprint = string.Empty;

    public LocalRpcCoreBridgeClient(CoreBridgeTransportOptions? options = null)
    {
        _opt = options ?? new CoreBridgeTransportOptions();
        _currentBackoffMs = _opt.PollIntervalMs;
        _http = new HttpClient
        {
            BaseAddress = new Uri(_opt.BaseUrl.TrimEnd('/') + "/"),
            Timeout = TimeSpan.FromMilliseconds(_opt.RequestTimeoutMs)
        };
    }

    public ConnectResponse Connect(ConnectRequest request)
    {
        var res = SendCommand<ConnectResponse>("connect", new { profile_id = request.ProfileId, mode = request.Mode }, "connect");
        return res ?? new ConnectResponse(string.Empty, "failed");
    }

    public DisconnectResponse Disconnect(DisconnectRequest request)
    {
        var res = SendCommand<DisconnectResponse>("disconnect", new { session_id = request.SessionId }, "disconnect");
        return res ?? new DisconnectResponse("failed");
    }

    public RankCandidatesResponse RankCandidates(string mode)
    {
        var res = SendCommand<RankCandidatesResponse>("rank_candidates", new { mode }, "rank_candidates");
        return res ?? new RankCandidatesResponse(Array.Empty<RankedCandidate>());
    }

    public FetchLogsResponse FetchLogs(string sessionId, string level)
    {
        var res = SendCommand<FetchLogsResponse>("fetch_logs", new { session_id = sessionId, level }, "fetch_logs");
        return res ?? new FetchLogsResponse(Array.Empty<LogEvent>());
    }

    public IDisposable Subscribe(Action<BridgeEvent> onEvent)
    {
        var id = Guid.NewGuid();
        _subs[id] = onEvent;
        EnsurePolling();

        // immediate health snapshot for new subscriber
        onEvent(new BridgeEvent("bridge_health_changed", new Dictionary<string, string>
        {
            ["healthy"] = _bridgeHealthy ? "true" : "false",
            ["reason"] = _lastBridgeReason
        }));

        return new Subscription(() =>
        {
            _subs.TryRemove(id, out _);
            if (_subs.IsEmpty)
            {
                StopPolling();
            }
        });
    }

    private T? SendCommand<T>(string command, object payload, string opName)
    {
        try
        {
            var body = JsonSerializer.Serialize(new RpcCommandRequest(command, payload));
            using var req = new StringContent(body, Encoding.UTF8, "application/json");
            using var resp = _http.PostAsync("command", req).GetAwaiter().GetResult();
            var raw = resp.Content.ReadAsStringAsync().GetAwaiter().GetResult();

            if (!resp.IsSuccessStatusCode)
            {
                MarkBridgeUnhealthy($"{opName}: http-{(int)resp.StatusCode}");
                return default;
            }

            var parsed = JsonSerializer.Deserialize<RpcCommandResponse<T>>(raw, JsonOpts());
            if (parsed is null || !parsed.Ok)
            {
                MarkBridgeUnhealthy($"{opName}: bad-response");
                return default;
            }

            MarkBridgeHealthy("ok");
            return parsed.Data;
        }
        catch (Exception ex)
        {
            MarkBridgeUnhealthy($"{opName}: {ex.GetType().Name}");
            return default;
        }
    }

    private void EnsurePolling()
    {
        if (_pollTask is { IsCompleted: false })
            return;

        _pollCts = new CancellationTokenSource();
        _pollTask = Task.Run(() => PollLoop(_pollCts.Token));
    }

    private async Task PollLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                using var resp = await _http.GetAsync("events/poll", ct);
                if (resp.IsSuccessStatusCode)
                {
                    var raw = await resp.Content.ReadAsStringAsync(ct);
                    var parsed = JsonSerializer.Deserialize<EventPollResponse>(raw, JsonOpts());
                    if (parsed is { Ok: true, Events: not null })
                    {
                        _pollFailureCount = 0;
                        _currentBackoffMs = _opt.PollIntervalMs;
                        MarkBridgeHealthy("ok");

                        foreach (var ev in parsed.Events)
                            BroadcastEventDedup(ev);
                    }
                    else
                    {
                        RegisterPollFailure("poll: invalid payload");
                    }
                }
                else
                {
                    RegisterPollFailure($"poll: http-{(int)resp.StatusCode}");
                }
            }
            catch (Exception ex)
            {
                RegisterPollFailure($"poll: {ex.GetType().Name}");
            }

            var delay = _pollFailureCount > 0 ? _currentBackoffMs : _opt.PollIntervalMs;
            try
            {
                await Task.Delay(delay, ct);
            }
            catch
            {
                return;
            }
        }
    }

    private void RegisterPollFailure(string reason)
    {
        _pollFailureCount++;
        if (_pollFailureCount >= _opt.FailureThresholdForDegraded)
            MarkBridgeUnhealthy(reason);

        var next = _currentBackoffMs <= 0 ? _opt.PollIntervalMs : _currentBackoffMs * 2;
        _currentBackoffMs = Math.Min(next, _opt.MaxBackoffMs);
    }

    private void MarkBridgeHealthy(string reason)
    {
        if (_bridgeHealthy && _lastBridgeReason == reason)
            return;

        _bridgeHealthy = true;
        _lastBridgeReason = reason;
        BroadcastEvent(new BridgeEvent("bridge_health_changed", new Dictionary<string, string>
        {
            ["healthy"] = "true",
            ["reason"] = reason
        }));
    }

    private void MarkBridgeUnhealthy(string reason)
    {
        if (!_bridgeHealthy && _lastBridgeReason == reason)
            return;

        _bridgeHealthy = false;
        _lastBridgeReason = reason;
        BroadcastEvent(new BridgeEvent("bridge_health_changed", new Dictionary<string, string>
        {
            ["healthy"] = "false",
            ["reason"] = reason
        }));
    }

    private void BroadcastEvent(BridgeEvent ev)
    {
        foreach (var sub in _subs.Values)
            sub(ev);
    }

    private void BroadcastEventDedup(BridgeEvent ev)
    {
        var fingerprint = $"{ev.Name}:{string.Join(';', ev.Fields.OrderBy(x => x.Key).Select(x => x.Key + "=" + x.Value))}";
        if (fingerprint == _lastEventFingerprint)
            return;
        _lastEventFingerprint = fingerprint;
        BroadcastEvent(ev);
    }

    private void StopPolling()
    {
        if (_pollCts == null)
            return;
        _pollCts.Cancel();
        _pollCts.Dispose();
        _pollCts = null;
        _pollTask = null;
        _pollFailureCount = 0;
        _currentBackoffMs = _opt.PollIntervalMs;
    }

    private static JsonSerializerOptions JsonOpts() => new()
    {
        PropertyNameCaseInsensitive = true
    };

    public void Dispose()
    {
        StopPolling();
        _http.Dispose();
    }

    private record RpcCommandRequest(string Command, object Payload);

    private record RpcCommandResponse<T>(bool Ok, T? Data, string? Error);

    private record EventPollResponse(bool Ok, List<BridgeEvent>? Events, string? Error);

    private class Subscription : IDisposable
    {
        private readonly Action _onDispose;
        private bool _disposed;

        public Subscription(Action onDispose) => _onDispose = onDispose;

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            _onDispose();
        }
    }
}
